worlds/pokemon_emerald/__init__.py

```diff
@@ -0,0 +1,841 @@
+                [
+                    ("TRAINER_BRENDAN_ROUTE_103_TREECKO", 0, False),
+                    ("TRAINER_BRENDAN_RUSTBORO_TREECKO",  1, False),
+                    ("TRAINER_BRENDAN_ROUTE_110_TREECKO", 2, True ),
+                    ("TRAINER_BRENDAN_ROUTE_119_TREECKO", 2, True ),
+                    ("TRAINER_BRENDAN_LILYCOVE_TREECKO",  3, True ),
+                    ("TRAINER_MAY_ROUTE_103_TREECKO",     0, False),
+                    ("TRAINER_MAY_RUSTBORO_TREECKO",      1, False),
+                    ("TRAINER_MAY_ROUTE_110_TREECKO",     2, True ),
+                    ("TRAINER_MAY_ROUTE_119_TREECKO",     2, True ),
+                    ("TRAINER_MAY_LILYCOVE_TREECKO",      3, True )
+                ],
+                [
+                    ("TRAINER_BRENDAN_ROUTE_103_TORCHIC", 0, False),
+                    ("TRAINER_BRENDAN_RUSTBORO_TORCHIC",  1, False),
+                    ("TRAINER_BRENDAN_ROUTE_110_TORCHIC", 2, True ),
+                    ("TRAINER_BRENDAN_ROUTE_119_TORCHIC", 2, True ),
+                    ("TRAINER_BRENDAN_LILYCOVE_TORCHIC",  3, True ),
+                    ("TRAINER_MAY_ROUTE_103_TORCHIC",     0, False),
+                    ("TRAINER_MAY_RUSTBORO_TORCHIC",      1, False),
+                    ("TRAINER_MAY_ROUTE_110_TORCHIC",     2, True ),
+                    ("TRAINER_MAY_ROUTE_119_TORCHIC",     2, True ),
+                    ("TRAINER_MAY_LILYCOVE_TORCHIC",      3, True )
+                ],
+                [
+                    ("TRAINER_BRENDAN_ROUTE_103_MUDKIP", 0, False),
+                    ("TRAINER_BRENDAN_RUSTBORO_MUDKIP",  1, False),
+                    ("TRAINER_BRENDAN_ROUTE_110_MUDKIP", 2, True ),
+                    ("TRAINER_BRENDAN_ROUTE_119_MUDKIP", 2, True ),
+                    ("TRAINER_BRENDAN_LILYCOVE_MUDKIP",  3, True ),
+                    ("TRAINER_MAY_ROUTE_103_MUDKIP",     0, False),
+                    ("TRAINER_MAY_RUSTBORO_MUDKIP",      1, False),
+                    ("TRAINER_MAY_ROUTE_110_MUDKIP",     2, True ),
+                    ("TRAINER_MAY_ROUTE_119_MUDKIP",     2, True ),
+                    ("TRAINER_MAY_LILYCOVE_MUDKIP",      3, True )
+                ]
+            ]
+
+            for i, starter in enumerate([new_starters[1], new_starters[2], new_starters[0]]):
+                potential_evolutions = [evolution.species_id for evolution in starter.evolutions]
+                picked_evolution = starter.species_id
+                if len(potential_evolutions) > 0:
+                    picked_evolution = self.random.choice(potential_evolutions)
+
+                for trainer_name, starter_position, is_evolved in rival_teams[i]:
+                    trainer_data = self.modified_data.trainers[emerald_data.constants[trainer_name]]
+                    trainer_data.party.pokemon[starter_position].species_id = picked_evolution if is_evolved else starter.species_id
+
+        self.modified_data = copy.deepcopy(emerald_data)
```

Comment:
Can all of this be assigned to a thread in post_fill and then awaited before you generate the output? progression balancing and accessibility check are run between those steps, so you wouldn't be able to do anything that relied on knowing exactly where something is.

Reply:
I suppose so. Nothing between the top of this function and the last line of this function currently cares about items or locations. But for that reason I'm not sure I see the purpose except that `post_fill` is conceptually for "randomizing things" while `generate_output` is conceptually for "creating a patch file". Especially if `generate_output` is already threaded.

Reply:
it's very much not necessary, i just like finding speed improvements. if you don't need to know any items or locations you could start the thread as early as `generate_basic` which i suspect would make a noticeable difference. but again, not necessary.

Reply:
Ah, I see. I'll keep it in mind and do some experiments later, but probably won't make those modifications in this PR.